### PR TITLE
WiP: separate introspection and call in operation

### PR DESCRIPTION
**Pull request by jcupitt**
The introspection part of Operation.call is now split out into a separate Introspect class, with a cache.

About 30% faster on the operation-call perf test, and it removes some code duplication.

It could be made a bit prettier, and perhaps faster.

### CODE REVIEW

**Comment by jcupitt** [comment]
It passes pytest, but I've not checked the doc generators.

**Comment by jcupitt** [comment]
I sped it up a bit more. Here's git master on `operation-call.py`:

```
  5th percentile: 123 us (-3% of the mean)
 25th percentile: 125 us (-2% of the mean) -- Q1
 50th percentile: 127 us (-0% of the mean) -- median
 75th percentile: 129 us (+1% of the mean) -- Q3
 95th percentile: 131 us (+3% of the mean)
```

And this branch is:

```
  5th percentile: 74.6 us (-3% of the mean)
 25th percentile: 75.3 us (-2% of the mean) -- Q1
 50th percentile: 76.7 us (-0% of the mean) -- median
 75th percentile: 78.3 us (+2% of the mean) -- Q3
 95th percentile: 81.1 us (+5% of the mean)
```

So almost twice as fast.

**Comment by kleisauke** [comment]
Nice, I also see a improvement of almost 30% on the operation-call perf test!

```bash
+----------------+--------------------+-----------------------------+
| Benchmark      | operation-call-old | operation-call-slots        |
+================+====================+=============================+
| Operation.call | 69.5 us            | 65.8 us: 1.06x faster (-5%) |
+----------------+--------------------+-----------------------------+
+----------------+---------------------------+---------------------------------+
| Benchmark      | operation-call-introspect | operation-call-introspect-slots |
+================+===========================+=================================+
| Operation.call | 52.2 us                   | 50.9 us: 1.03x faster (-3%)     |
+----------------+---------------------------+---------------------------------+
+----------------+--------------------+---------------------------------+
| Benchmark      | operation-call-old | operation-call-introspect-slots |
+================+====================+=================================+
| Operation.call | 69.5 us            | 50.9 us: 1.36x faster (-27%)    |
+----------------+--------------------+---------------------------------+
```

So the `__slots__` patch saves about 3% - 5%:
<details><summary>__slots__ patch</summary>

```diff
From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sun, 17 Nov 2019 12:20:00 +0100
Subject: [PATCH] Use the special __slots__ attribute

For faster attribute access and to reduce the memory footprint.

diff --git a/pyvips/gobject.py b/pyvips/gobject.py
index 1111111..2222222 100644
--- a/pyvips/gobject.py
+++ b/pyvips/gobject.py
@@ -41,6 +41,7 @@ class GObject(object):
     """Manage GObject lifetime.
 
     """
+    __slots__ = ('_handles', 'pointer')
 
     def __init__(self, pointer):
         """Wrap around a pointer.
@@ -55,12 +56,9 @@ class GObject(object):
         self._handles = []
 
         # record the pointer we were given to manage
-        self.pointer = pointer
-        # logger.debug('GObject.__init__: pointer = %s', str(self.pointer))
-
         # on GC, unref
-        self.gobject = ffi.gc(self.pointer, gobject_lib.g_object_unref)
-        # logger.debug('GObject.__init__: gobject = %s', str(self.gobject))
+        self.pointer = ffi.gc(pointer, gobject_lib.g_object_unref)
+        # logger.debug('GObject.__init__: pointer = %s', str(self.pointer))
 
     def signal_connect(self, name, callback):
         """Connect to a signal on this object.
diff --git a/pyvips/gvalue.py b/pyvips/gvalue.py
index 1111111..2222222 100644
--- a/pyvips/gvalue.py
+++ b/pyvips/gvalue.py
@@ -29,6 +29,7 @@ class GValue(object):
     GValue lifetime is managed automatically.
 
     """
+    __slots__ = ('pointer', 'gvalue')
 
     # look up some common gtypes at init for speed
     gbool_type = type_from_name('gboolean')
diff --git a/pyvips/vimage.py b/pyvips/vimage.py
index 1111111..2222222 100644
--- a/pyvips/vimage.py
+++ b/pyvips/vimage.py
@@ -83,16 +83,23 @@ def _run_cmplx(fn, image):
 
 
 # https://stackoverflow.com/a/22409540/1480019
-def _with_metaclass(mcls):
-    def decorator(cls):
-        body = vars(cls).copy()
-        # clean out class body
-        body.pop('__dict__', None)
-        body.pop('__weakref__', None)
-
-        return mcls(cls.__name__, cls.__bases__, body)
-
-    return decorator
+# https://github.com/benjaminp/six/blob/33b584b2c551548021adb92a028ceaf892deb5be/six.py#L846-L861
+def _with_metaclass(metaclass):
+    """Class decorator for creating a class with a metaclass."""
+    def wrapper(cls):
+        orig_vars = cls.__dict__.copy()
+        slots = orig_vars.get('__slots__')
+        if slots is not None:
+            if isinstance(slots, str):
+                slots = [slots]
+            for slots_var in slots:
+                orig_vars.pop(slots_var)
+        orig_vars.pop('__dict__', None)
+        orig_vars.pop('__weakref__', None)
+        if hasattr(cls, '__qualname__'):
+            orig_vars['__qualname__'] = cls.__qualname__
+        return metaclass(cls.__name__, cls.__bases__, orig_vars)
+    return wrapper
 
 
 # decorator to set docstring
@@ -135,6 +142,7 @@ class Image(pyvips.VipsObject):
     """Wrap a VipsImage object.
 
     """
+    __slots__ = ('_references',)
 
     # private static
 
diff --git a/pyvips/vobject.py b/pyvips/vobject.py
index 1111111..2222222 100644
--- a/pyvips/vobject.py
+++ b/pyvips/vobject.py
@@ -13,10 +13,12 @@ logger = logging.getLogger(__name__)
 
 class VipsObject(pyvips.GObject):
     """Manage a VipsObject."""
+    __slots__ = ('vobject',)
 
     def __init__(self, pointer):
         # logger.debug('VipsObject.__init__: pointer = %s', pointer)
         super(VipsObject, self).__init__(pointer)
+        self.vobject = ffi.cast('VipsObject*', pointer)
 
     @staticmethod
     def print_all(msg):
@@ -39,8 +41,7 @@ class VipsObject(pyvips.GObject):
         pspec = ffi.new('GParamSpec **')
         argument_class = ffi.new('VipsArgumentClass **')
         argument_instance = ffi.new('VipsArgumentInstance **')
-        vo = ffi.cast('VipsObject *', self.pointer)
-        result = vips_lib.vips_object_get_argument(vo, _to_bytes(name),
+        result = vips_lib.vips_object_get_argument(self.vobject, _to_bytes(name),
                                                    pspec, argument_class,
                                                    argument_instance)
 
@@ -121,17 +122,15 @@ class VipsObject(pyvips.GObject):
 
         """
 
-        vo = ffi.cast('VipsObject *', self.pointer)
         cstr = _to_bytes(string_options)
-        result = vips_lib.vips_object_set_from_string(vo, cstr)
+        result = vips_lib.vips_object_set_from_string(self.vobject, cstr)
 
         return result == 0
 
     def get_description(self):
         """Get the description of a GObject."""
 
-        vo = ffi.cast('VipsObject *', self.pointer)
-        return _to_string(vips_lib.vips_object_get_description(vo))
+        return _to_string(vips_lib.vips_object_get_description(self.vobject))
 
 
 __all__ = ['VipsObject']
diff --git a/pyvips/voperation.py b/pyvips/voperation.py
index 1111111..2222222 100644
--- a/pyvips/voperation.py
+++ b/pyvips/voperation.py
@@ -51,7 +51,7 @@ class Introspect(object):
             p_names = ffi.new('char**[1]')
             p_flags = ffi.new('int*[1]')
             p_n_args = ffi.new('int[1]')
-            result = vips_lib.vips_object_get_args(op.object,
+            result = vips_lib.vips_object_get_args(op.vobject,
                                                    p_names, p_flags, p_n_args)
             if result != 0:
                 raise Error('unable to get arguments from operation')
@@ -68,7 +68,7 @@ class Introspect(object):
                 return ffi.NULL
 
             cb = ffi.callback('VipsArgumentMapFn', add_construct)
-            vips_lib.vips_argument_map(self.object, cb, ffi.NULL, ffi.NULL)
+            vips_lib.vips_argument_map(op.vobject, cb, ffi.NULL, ffi.NULL)
 
         # logger.debug('arguments = %s', self.arguments)
 
@@ -162,6 +162,7 @@ class Operation(pyvips.VipsObject):
     This class wraps the libvips VipsOperation class.
 
     """
+    __slots__ = ()
 
     # cache nickname -> docstring here
     _docstring_cache = {}
@@ -169,7 +170,6 @@ class Operation(pyvips.VipsObject):
     def __init__(self, pointer):
         # logger.debug('Operation.__init__: pointer = %s', pointer)
         super(Operation, self).__init__(pointer)
-        self.object = ffi.cast('VipsObject*', pointer)
 
     @staticmethod
     def new_from_name(operation_name):
@@ -303,7 +303,7 @@ class Operation(pyvips.VipsObject):
         if len(opts) > 0:
             result.append(opts)
 
-        vips_lib.vips_object_unref_outputs(op.object)
+        vips_lib.vips_object_unref_outputs(op.vobject)
 
         if len(result) == 0:
             result = None

```
</details>

Here are the full results:
<details><summary>Full results</summary>

```bash
$ python3 -m pyperf stats operation-call-slots.json
Total duration: 11.8 sec
Start date: 2019-11-17 12:02:17
End date: 2019-11-17 12:02:30
Raw value minimum: 132 ms
Raw value maximum: 138 ms

Number of calibration run: 1
Number of run with values: 20
Total number of run: 21

Number of warmup per run: 1
Number of value per run: 3
Loop iterations per value: 2048
Total number of values: 60

Minimum:         64.4 us
Median +- MAD:   65.8 us +- 0.5 us
Mean +- std dev: 65.8 us +- 0.7 us
Maximum:         67.3 us

  0th percentile: 64.4 us (-2% of the mean) -- minimum
  5th percentile: 64.5 us (-2% of the mean)
 25th percentile: 65.3 us (-1% of the mean) -- Q1
 50th percentile: 65.8 us (+0% of the mean) -- median
 75th percentile: 66.3 us (+1% of the mean) -- Q3
 95th percentile: 66.8 us (+2% of the mean)
100th percentile: 67.3 us (+2% of the mean) -- maximum

Number of outlier (out of 63.7 us..67.9 us): 0
$ python3 -m pyperf stats operation-call-old.json
Total duration: 12.4 sec
Start date: 2019-11-17 12:03:12
End date: 2019-11-17 12:03:26
Raw value minimum: 140 ms
Raw value maximum: 146 ms

Number of calibration run: 1
Number of run with values: 20
Total number of run: 21

Number of warmup per run: 1
Number of value per run: 3
Loop iterations per value: 2048
Total number of values: 60

Minimum:         68.3 us
Median +- MAD:   69.3 us +- 0.6 us
Mean +- std dev: 69.5 us +- 0.7 us
Maximum:         71.1 us

  0th percentile: 68.3 us (-2% of the mean) -- minimum
  5th percentile: 68.5 us (-1% of the mean)
 25th percentile: 68.8 us (-1% of the mean) -- Q1
 50th percentile: 69.3 us (-0% of the mean) -- median
 75th percentile: 70.0 us (+1% of the mean) -- Q3
 95th percentile: 70.7 us (+2% of the mean)
100th percentile: 71.1 us (+2% of the mean) -- maximum

Number of outlier (out of 66.9 us..71.9 us): 0
$ python3 -m pyperf stats operation-call-introspect-slots.json
Total duration: 9.2 sec
Start date: 2019-11-17 12:11:38
End date: 2019-11-17 12:11:49
Raw value minimum: 101 ms
Raw value maximum: 108 ms

Number of calibration run: 1
Number of run with values: 20
Total number of run: 21

Number of warmup per run: 1
Number of value per run: 3
Loop iterations per value: 2048
Total number of values: 60

Minimum:         49.4 us
Median +- MAD:   50.9 us +- 0.4 us
Mean +- std dev: 50.9 us +- 0.8 us
Maximum:         52.8 us

  0th percentile: 49.4 us (-3% of the mean) -- minimum
  5th percentile: 49.5 us (-3% of the mean)
 25th percentile: 50.5 us (-1% of the mean) -- Q1
 50th percentile: 50.9 us (+0% of the mean) -- median
 75th percentile: 51.3 us (+1% of the mean) -- Q3
 95th percentile: 52.5 us (+3% of the mean)
100th percentile: 52.8 us (+4% of the mean) -- maximum

Number of outlier (out of 49.4 us..52.5 us): 4
$ python3 -m pyperf stats operation-call-introspect.json
Total duration: 9.4 sec
Start date: 2019-11-17 12:13:28
End date: 2019-11-17 12:13:40
Raw value minimum: 103 ms
Raw value maximum: 110 ms

Number of calibration run: 1
Number of run with values: 20
Total number of run: 21

Number of warmup per run: 1
Number of value per run: 3
Loop iterations per value: 2048
Total number of values: 60

Minimum:         50.3 us
Median +- MAD:   52.3 us +- 0.6 us
Mean +- std dev: 52.2 us +- 0.8 us
Maximum:         53.8 us

  0th percentile: 50.3 us (-4% of the mean) -- minimum
  5th percentile: 51.3 us (-2% of the mean)
 25th percentile: 51.5 us (-1% of the mean) -- Q1
 50th percentile: 52.3 us (+0% of the mean) -- median
 75th percentile: 52.8 us (+1% of the mean) -- Q3
 95th percentile: 53.5 us (+2% of the mean)
100th percentile: 53.8 us (+3% of the mean) -- maximum

Number of outlier (out of 49.6 us..54.8 us): 0
```
</details>

**Comment by jcupitt** [comment]
Nice! I pasted in your slots patch.

Do you think this is OK to merge?

File: pyvips/voperation.py [hunk 1]
```diff
@@ -22,6 +22,124 @@
...
+class Introspect(object):
+    """Build introspection data for operations.
+
+    Make an operation, introspect it, and build a structure representing
+    everything we know about it.
+
+    """
+
+    def __init__(self, operation_name):
+        op = Operation.new_from_name(operation_name)
```
Comment (kleisauke) [inline pyvips/voperation.py]: We're constructing a new operation now twice, see:
https://github.com/libvips/pyvips/blob/5f7c5d3c823497d494141e0f900562824ed12edb/pyvips/voperation.py#L220-L228

Perhaps we should only do this once?

**Comment by kleisauke** [comment]
This is OK to merge. I only noticed that `Operation.new_from_name(operation_name)` is called twice within `Operation.call`, but maybe this doesn't matter much.

fwiw, I've a few small performance patches left:
<details><summary>Patches</summary>

```diff
From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Mon, 18 Nov 2019 12:44:37 +0100
Subject: [PATCH 1/3] Cast to VipsObject within the constructor

Avoids casting in several places.

diff --git a/pyvips/gobject.py b/pyvips/gobject.py
index 1111111..2222222 100644
--- a/pyvips/gobject.py
+++ b/pyvips/gobject.py
@@ -41,7 +41,7 @@ class GObject(object):
     """Manage GObject lifetime.
 
     """
-    __slots__ = '_handles', 'pointer', 'gobject'
+    __slots__ = ('_handles', 'pointer', 'gobject')
 
     def __init__(self, pointer):
         """Wrap around a pointer.
diff --git a/pyvips/gvalue.py b/pyvips/gvalue.py
index 1111111..2222222 100644
--- a/pyvips/gvalue.py
+++ b/pyvips/gvalue.py
@@ -29,7 +29,7 @@ class GValue(object):
     GValue lifetime is managed automatically.
 
     """
-    __slots__ = 'pointer', 'gvalue'
+    __slots__ = ('pointer', 'gvalue')
 
     # look up some common gtypes at init for speed
     gbool_type = type_from_name('gboolean')
diff --git a/pyvips/vimage.py b/pyvips/vimage.py
index 1111111..2222222 100644
--- a/pyvips/vimage.py
+++ b/pyvips/vimage.py
@@ -142,7 +142,7 @@ class Image(pyvips.VipsObject):
     """Wrap a VipsImage object.
 
     """
-    __slots__ = '_references',
+    __slots__ = ('_references',)
 
     # private static
 
diff --git a/pyvips/vobject.py b/pyvips/vobject.py
index 1111111..2222222 100644
--- a/pyvips/vobject.py
+++ b/pyvips/vobject.py
@@ -13,10 +13,12 @@ logger = logging.getLogger(__name__)
 
 class VipsObject(pyvips.GObject):
     """Manage a VipsObject."""
+    __slots__ = ('vobject',)
 
     def __init__(self, pointer):
         # logger.debug('VipsObject.__init__: pointer = %s', pointer)
         super(VipsObject, self).__init__(pointer)
+        self.vobject = ffi.cast('VipsObject*', pointer)
 
     @staticmethod
     def print_all(msg):
@@ -39,8 +41,7 @@ class VipsObject(pyvips.GObject):
         pspec = ffi.new('GParamSpec **')
         argument_class = ffi.new('VipsArgumentClass **')
         argument_instance = ffi.new('VipsArgumentInstance **')
-        vo = ffi.cast('VipsObject *', self.pointer)
-        result = vips_lib.vips_object_get_argument(vo, _to_bytes(name),
+        result = vips_lib.vips_object_get_argument(self.vobject, _to_bytes(name),
                                                    pspec, argument_class,
                                                    argument_instance)
 
@@ -121,17 +122,15 @@ class VipsObject(pyvips.GObject):
 
         """
 
-        vo = ffi.cast('VipsObject *', self.pointer)
         cstr = _to_bytes(string_options)
-        result = vips_lib.vips_object_set_from_string(vo, cstr)
+        result = vips_lib.vips_object_set_from_string(self.vobject, cstr)
 
         return result == 0
 
     def get_description(self):
         """Get the description of a GObject."""
 
-        vo = ffi.cast('VipsObject *', self.pointer)
-        return _to_string(vips_lib.vips_object_get_description(vo))
+        return _to_string(vips_lib.vips_object_get_description(self.vobject))
 
 
 __all__ = ['VipsObject']
diff --git a/pyvips/voperation.py b/pyvips/voperation.py
index 1111111..2222222 100644
--- a/pyvips/voperation.py
+++ b/pyvips/voperation.py
@@ -51,7 +51,7 @@ class Introspect(object):
             p_names = ffi.new('char**[1]')
             p_flags = ffi.new('int*[1]')
             p_n_args = ffi.new('int[1]')
-            result = vips_lib.vips_object_get_args(op.object,
+            result = vips_lib.vips_object_get_args(op.vobject,
                                                    p_names, p_flags, p_n_args)
             if result != 0:
                 raise Error('unable to get arguments from operation')
@@ -68,7 +68,7 @@ class Introspect(object):
                 return ffi.NULL
 
             cb = ffi.callback('VipsArgumentMapFn', add_construct)
-            vips_lib.vips_argument_map(self.object, cb, ffi.NULL, ffi.NULL)
+            vips_lib.vips_argument_map(op.vobject, cb, ffi.NULL, ffi.NULL)
 
         # logger.debug('arguments = %s', self.arguments)
 
@@ -162,7 +162,7 @@ class Operation(pyvips.VipsObject):
     This class wraps the libvips VipsOperation class.
 
     """
-    __slots__ = 'object',
+    __slots__ = ()
 
     # cache nickname -> docstring here
     _docstring_cache = {}
@@ -170,7 +170,6 @@ class Operation(pyvips.VipsObject):
     def __init__(self, pointer):
         # logger.debug('Operation.__init__: pointer = %s', pointer)
         super(Operation, self).__init__(pointer)
-        self.object = ffi.cast('VipsObject*', pointer)
 
     @staticmethod
     def new_from_name(operation_name):
@@ -304,7 +303,7 @@ class Operation(pyvips.VipsObject):
         if len(opts) > 0:
             result.append(opts)
 
-        vips_lib.vips_object_unref_outputs(op.object)
+        vips_lib.vips_object_unref_outputs(op.vobject)
 
         if len(result) == 0:
             result = None

From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Mon, 18 Nov 2019 12:47:17 +0100
Subject: [PATCH 2/3] Implement __slots__ within the Introspect class


diff --git a/pyvips/voperation.py b/pyvips/voperation.py
index 1111111..2222222 100644
--- a/pyvips/voperation.py
+++ b/pyvips/voperation.py
@@ -29,23 +29,26 @@ class Introspect(object):
     everything we know about it.
 
     """
+    __slots__ = ('description', 'flags', 'details',
+                 'required_input', 'optional_input',
+                 'required_output', 'optional_output',
+                 'member_x', 'method_args')
 
     def __init__(self, operation_name):
         op = Operation.new_from_name(operation_name)
 
-        self.name = operation_name
         self.description = op.get_description()
         self.flags = vips_lib.vips_operation_get_flags(op.pointer)
 
         # build a list of constructor arg [name, flags] pairs in arg order
-        self.arguments = []
+        arguments = []
 
         def add_args(name, flags):
             if (flags & _CONSTRUCT) != 0:
                 # libvips uses '-' to separate parts of arg names, but we
                 # need '_' for Python
                 name = name.replace('-', '_')
-                self.arguments.append([name, flags])
+                arguments.append([name, flags])
 
         if at_least_libvips(8, 7):
             p_names = ffi.new('char**[1]')
@@ -74,7 +77,7 @@ class Introspect(object):
 
         # build a hash from arg name to detailed arg information
         self.details = {}
-        for name, flags in self.arguments:
+        for name, flags in arguments:
             self.details[name] = {
                 "name": name,
                 "flags": flags,
@@ -88,7 +91,7 @@ class Introspect(object):
         self.required_output = []
         self.optional_output = []
 
-        for name, flags in self.arguments:
+        for name, flags in arguments:
             if ((flags & _INPUT) != 0 and
                     (flags & _REQUIRED) != 0 and
                     (flags & _DEPRECATED) == 0):

From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Mon, 18 Nov 2019 12:57:05 +0100
Subject: [PATCH 3/3] Garbage-collect the pointer within the same variable

ffi.gc does not need to be in a separate variable.

diff --git a/pyvips/gobject.py b/pyvips/gobject.py
index 1111111..2222222 100644
--- a/pyvips/gobject.py
+++ b/pyvips/gobject.py
@@ -41,7 +41,7 @@ class GObject(object):
     """Manage GObject lifetime.
 
     """
-    __slots__ = ('_handles', 'pointer', 'gobject')
+    __slots__ = ('_handles', 'pointer')
 
     def __init__(self, pointer):
         """Wrap around a pointer.
@@ -56,12 +56,9 @@ class GObject(object):
         self._handles = []
 
         # record the pointer we were given to manage
-        self.pointer = pointer
-        # logger.debug('GObject.__init__: pointer = %s', str(self.pointer))
-
         # on GC, unref
-        self.gobject = ffi.gc(self.pointer, gobject_lib.g_object_unref)
-        # logger.debug('GObject.__init__: gobject = %s', str(self.gobject))
+        self.pointer = ffi.gc(pointer, gobject_lib.g_object_unref)
+        # logger.debug('GObject.__init__: pointer = %s', str(self.pointer))
 
     def signal_connect(self, name, callback):
         """Connect to a signal on this object.
```
</details>

If you want, I can open a separate PR when this one lands.

File: pyvips/voperation.py (hunk 1 above)
Comment (jcupitt) [reply pyvips/voperation.py]: The one in `Introspect` is only made once (after that, intro data is fetched from the cache), so I don't think it matters.

**Comment by jcupitt** [comment]
Looks reasonable -- committed!

**Pull request merged**